python/pip_install/repositories.bzl

```diff
@@ -74,11 +73,11 @@ def pip_install_dependencies():
     versions.check("4.0.0")
 
     for (name, url, sha256) in _RULE_DEPS:
-        maybe(
-            http_archive,
-            name,
-            url = url,
-            sha256 = sha256,
-            type = "zip",
-            build_file_content = _GENERIC_WHEEL,
-        )
+        if not native.existing_rule(name):
```

Comment:
FWIW bazel recommends not using native.existing_rule.
>If possible, avoid using this function. It makes BUILD files brittle and order-dependent.

Reply:
Also

> [Note that the existing_rule(s) functions are poorly defined, have design issues, can lead to performance and maintenance issues. So we don't really want to encourage their use.](https://github.com/bazelbuild/bazel/issues/7496)


It seems `rules_go` has for many years used `native.existing_rules` to have a 'fake `maybe`' https://github.com/bazelbuild/rules_go/pull/782/files.

Reply:
I think that note is describing common issues in WORKSPACE management. I don't understand why `maybe` would be ok but not the native functions. One argument might be because `maybe` can't be used recursively but I think this situation is notably different in that users should be able to treat the `pip_repository` rules as normal repository rules, but there's some macro magic involved to inject the necessary dependencies. I think the use of these native functions here is fine but I'm also happy to learn better practices 😄